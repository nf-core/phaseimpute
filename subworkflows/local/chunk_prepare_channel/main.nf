workflow CHUNK_PREPARE_CHANNEL {

    take:
    ch_chunks // channel:   [ [id, chr], txt ]
    tool

    main:

    if(tool == "glimpse"){
        ch_chunks = ch_chunks.map { chr, txt -> [chr, file(txt)]}
                .splitCsv(header: ['ID', 'Chr', 'RegionIn', 'RegionOut', 'Size1', 'Size2'], sep: "\t", skip: 0)
                .map { meta, it -> [meta, it["RegionIn"], it["RegionOut"]]}
    } else if(tool == "quilt") {
        ch_chunks = ch_chunks.map { chr, txt -> [chr, file(txt)]}
            .splitCsv(header: ['ID', 'Chr', 'RegionIn', 'RegionOut', 'Size1', 'Size2'], sep: "\t", skip: 0)
            .map { meta, it ->
                def startEnd = it["RegionIn"].split(':')[1].split('-')
                [meta, meta.chr, startEnd[0], startEnd[1]]
            }
    } else {
        error "Only 'glimpse' and 'quilt' output format are supported. Got ${tool}"
    }

    emit:
    chunks  = ch_chunks // channel:   [ [meta], regionstart, regionend ]

}
