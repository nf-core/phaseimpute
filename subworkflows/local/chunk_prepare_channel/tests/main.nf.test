nextflow_workflow {

    name "Test Subworkflow CHUNK_PREPARE_CHANNEL"
    script "../main.nf"

    workflow "CHUNK_PREPARE_CHANNEL"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/chunk_prepare_channel"
    tag "chunk_prepare_channel"

    test("Prepare chunks channel Glimpse") {
        when {
            params {
                max_cpus   = 2
                max_memory = '2.GB'
            }
            workflow {
                """
                input[0] = Channel.of([
                    [id: "1000GP", chr: "chr21"],
                    file(params.pipelines_testdata_base_path + "hum_data/panel/chr21/1000GP.chr21_chunks.txt", checkIfExist:true),
                ])
                input[1] = "glimpse"
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
            )
        }
    }

    test("Prepare chunks channel Quilt") {
        when {
            params {
                max_cpus   = 2
                max_memory = '2.GB'
            }
            workflow {
                """
                input[0] = Channel.of([
                    [id: "1000GP", chr: "chr21"],
                    file(params.pipelines_testdata_base_path + "hum_data/panel/chr21/1000GP.chr21_chunks.txt", checkIfExist:true),
                ])
                input[1] = "quilt"
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
            )
        }
    }

    test("Prepare chunks channel error tool") {
        when {
            params {
                max_cpus   = 2
                max_memory = '2.GB'
            }
            workflow {
                """
                input[0] = Channel.of([
                    [id: "1000GP", chr: "chr21"],
                    file(params.pipelines_testdata_base_path + "hum_data/panel/chr21/1000GP.chr21_chunks.txt", checkIfExist:true),
                ])
                input[1] = "glimpse2"
                """
            }
        }

        then {
            assertAll(
                { assert workflow.failed },
                { assert snapshot(workflow.stdout).match() }
            )
        }
    }
}
