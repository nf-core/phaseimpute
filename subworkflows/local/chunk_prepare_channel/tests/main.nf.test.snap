{
    "Prepare chunks channel error tool": {
        "content": [
            [
                "N E X T F L O W  ~  version 23.10.1",
                "Launching `/mnt/c/Users/llenezet/Documents/Repositories/phaseimpute/.nf-test/tests/c7782a704e2a438fa20d101b0208a35d/meta/mock.nf` [hungry_wescoff] DSL2 - revision: c41c2d7547",
                "Only 'glimpse' and 'quilt' output format are supported. Got glimpse2"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-07-18T09:57:16.071041128"
    },
    "Prepare chunks channel Quilt": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr21"
                        },
                        "chr21",
                        "16570070",
                        "16595525"
                    ],
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr21"
                        },
                        "chr21",
                        "16585483",
                        "16609998"
                    ]
                ],
                "chunks": [
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr21"
                        },
                        "chr21",
                        "16570070",
                        "16595525"
                    ],
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr21"
                        },
                        "chr21",
                        "16585483",
                        "16609998"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-07-18T09:57:11.37570123"
    },
    "Prepare chunks channel Glimpse": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr21"
                        },
                        "chr21:16570070-16595525",
                        "chr21:16570070-16590513"
                    ],
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr21"
                        },
                        "chr21:16585483-16609998",
                        "chr21:16590521-16609998"
                    ]
                ],
                "chunks": [
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr21"
                        },
                        "chr21:16570070-16595525",
                        "chr21:16570070-16590513"
                    ],
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr21"
                        },
                        "chr21:16585483-16609998",
                        "chr21:16590521-16609998"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-07-18T09:57:06.040215496"
    }
}